File: Fall 2020/Programming Paradigms in C++ brooklyn college (CISC 3142)/Lab 4.1/rational.cpp
#include <iostream>
#include <cmath> 
using namespace std;
#include "rational.h"
#include "rational_exception.h"

namespace rational{
	int gcd(int numerator, int denominator) {
	    if (denominator == 0)
	        return numerator;
	    else
	        return gcd(denominator, numerator%denominator);
		}
}

Rational::Rational(int numerator, int denominator){
    if (denominator == 0)
        throw RationalException("Denominator can't be zero!");
    this->num = numerator/rational::gcd(numerator,denominator);
    this->denom = denominator/rational::gcd(numerator,denominator);
}

Rational Rational::neg() const{
    return Rational(-1*this->num,this->denom);
}

Rational Rational::inv() const{
    return Rational(this->denom,this->num);
}

Rational Rational::add(const Rational &operande) const{
    return Rational(
        this->num* operande.denom + operande.num * this->denom,
        this->denom * operande.denom
        );
}

Rational Rational::sub(const Rational &operande) const{
    Rational operandNegated = operande.neg();
    return this->add(operandNegated);
}

Rational Rational::mul(const Rational &operande) const{
    return Rational(this->num * operande.num, this->denom * operande.denom);
}

Rational Rational::div(const Rational &operande) const{
    Rational operandeInv = operande.inv();
    return this->mul(operandeInv);
}

int Rational::getNumerator() const{
    return this->num;
}

int Rational::getDenominator() const{
    return this->denom;
}

Rational &Rational::addInPlace(const Rational &operade){
    Rational sum = this->add(operade);
    this->num = sum.num;
    this->denom = sum.denom;
    return *this;
}

Rational &Rational::subInPlace(const Rational &operade){
    Rational difference = this->sub(operade);
    this->num = difference.num;
    this->denom = difference.denom;
    return *this;
}

Rational &Rational::mulInPlace(const Rational &operade){
    Rational product = this->sub(operade);
    this->num = product.num;
    this->denom = product.denom;
    return *this;
}

Rational &Rational::divInPlace(const Rational &operade){
    Rational quotient = this->sub(operade);
    this->num = quotient.num;
    this->denom = quotient.denom;
    return *this;
}

void Rational::print(ostream &os) const{
    if (this->denom == 1) 
        os << this->num;
    else if (this->num == this->denom)
        os << 1;
    else 
        os << this->num << "/" << this->denom;
}

int Rational::compareTo(const Rational &operande) const{
    float a = (float)this->getNumerator()/(float)this->getDenominator();
    float b = (float)operande.getNumerator()/(float)operande.getDenominator();
    if (a < b)
        return -1;
    else if (abs(a-b) <= 0.0001) 
        return 0;
    else
        return 1;
}

bool Rational::equals(const Rational &operande) const{
    return (this->compareTo(operande) == 0);
}

File: Fall 2020/Programming Paradigms in C++ brooklyn college (CISC 3142)/Lab 2.3/Rational.java
public class Rational{
	
	public static void main (String args) {
		Person m = new Person (new Name("mohamed", "Shafee"), 27);
		System.out.println(Person.x);
	}
	private int numerator, denominator;
	
	Rational(){
		this(0);
	}
	
	Rational(int numerator){
		this(numerator, 1);
	}
	
	Rational(Rational rational){
		this(rational.getNumerator(),rational.getDenominator());
	}
	
	Rational(int numerator, int denominator){
		if (denominator == 0) {
            throw new RationalException("Denominator can't be zero!");
        }
		this.numerator = numerator/gcd(numerator,denominator);
		this.denominator = denominator/gcd(numerator,denominator);
	}
	
	public Rational negate() {
		return new Rational(-1*this.numerator, this.denominator);
	}
	
	public Rational inverse() {
		return new Rational(this.denominator, this.numerator);
	}
	
	public Rational add( Rational operande) {
		return new Rational(
				this.numerator* operande.denominator + operande.getNumerator()*this.getDenominator(),
				this.denominator*operande.denominator
				);
	}
	
	public Rational sub(Rational operande){
		return this.add(operande.negate());
	}
	
	public Rational mul( Rational operande){
		return new Rational(this.numerator*operande.getNumerator(), this.denominator*operande.getDenominator());
	}
	
	public Rational div( Rational operande){
		return this.mul(operande.inverse());
	}
	
	public int getNumerator() {
		return this.numerator;
	}
	
	public int getDenominator() {
		return this.denominator;
	}
	
	private int gcd(int numerator, int denominator) {
		if (denominator == 0)
			return numerator;
		else
			return gcd(denominator, numerator%denominator);
	}
	
	public void addInPlace(Rational operande) {
		Rational temp = this.add(operande);
		this.numerator = temp.getNumerator();
		this.denominator = temp.getDenominator();
	}
	
	public void subInPlace(Rational operande) {
		Rational temp = this.sub(operande);
		this.numerator = temp.getNumerator();
		this.denominator = temp.getDenominator();
	}
	
	public void divInPlace( Rational operande) {
		Rational temp = this.div(operande);
		this.numerator = temp.getNumerator();
		this.denominator = temp.getDenominator();
	}
	
	public void mulInPlace( Rational operande) {
		Rational temp = this.mul(operande);
		this.numerator = temp.getNumerator();
		this.denominator = temp.getDenominator();
	}
	
	public int compareTo(Rational operande) {
		float a = (float)this.getNumerator()/(float)this.getDenominator();
		float b = (float)operande.getNumerator()/(float)operande.getDenominator();
		if (a < b)
			return -1;
		else if (Math.abs(a-b) <= 0.0001) 
			return 0;
		else
			return 1;
	}
	
	public boolean equals(Rational operande) {
		return (this.compareTo(operande) == 0);
	}
	
	public String toString() {
		return this.denominator == 1 ? Integer.toString(this.numerator) :
		this.numerator == this.denominator? "1":
		Integer.toString(this.numerator) + "/" + Integer.toString(this.denominator);
	}
}